perun-core/src/test/java/cz/metacentrum/perun/core/entry/MembersManagerEntryIntegrationTest.java

```diff
@@ -2207,6 +2208,70 @@ public void getAllSponsoredMembers() throws Exception {
+		User targetUser = perun.getUsersManagerBl().getUserByMember(sess, member);
+		// delete member - member was created just to create target user
+		perun.getMembersManager().deleteMember(sess, member);
+		// save user for deletion after test
+		usersForDeletion.add(targetUser);
+
+		// add createdMember to createdGroup
+		groupsManagerEntry.addMember(sess, createdGroup, createdMember);
+
+		// set attributes
+		Attribute memberAttribute = setUpAttribute(Integer.class.getName(), "testMemberAttribute", AttributesManager.NS_MEMBER_ATTR_DEF, 15);
+		perun.getAttributesManagerBl().setAttribute(sess, createdMember, memberAttribute);
+		Attribute memberGroupAttribute = setUpAttribute(Integer.class.getName(), "testMemberGroupAttribute", AttributesManager.NS_MEMBER_GROUP_ATTR_DEF, 15);
+		perun.getAttributesManagerBl().setAttribute(sess, createdMember, createdGroup, memberGroupAttribute);
+
+		List<Attribute> sourceMemberAttrs = perun.getAttributesManagerBl().getAttributes(sess, createdMember);
+		List<Attribute> sourceMemberGroupAttrs = perun.getAttributesManagerBl().getAttributes(sess, createdMember, createdGroup);
+
+		BanOnVo banOnVo = new BanOnVo(-1, createdMember.getId(), createdMember.getVoId(), null, "ban");
+		perun.getVosManagerBl().setBan(sess, banOnVo);
+
+		perun.getMembersManager().moveMembership(sess, createdVo, sourceUser, targetUser);
+
+		Member targetMember = perun.getMembersManagerBl().getMemberByUser(sess, createdVo, targetUser);
+		List<Attribute> targetMemberAttrs = perun.getAttributesManagerBl().getAttributes(sess, targetMember);
+		List<Attribute> targetMemberGroupAttrs = perun.getAttributesManagerBl().getAttributes(sess, targetMember, createdGroup);
+
+		assertEquals("The source and target member attributes should have the same size.",
+			sourceMemberAttrs.size(), targetMemberAttrs.size());
+		assertEquals("The source and target member-group attributes should have the same size.",
+			sourceMemberGroupAttrs.size(), targetMemberGroupAttrs.size());
+
+		// remove ID attribute from list because it is the only different attribute and then compare lists
+		targetMemberAttrs.removeIf(attribute -> attribute.getName().equals(AttributesManager.NS_MEMBER_ATTR_CORE + ":id"));
+
+		assertTrue("The source and target member should have the same member attributes.",
+			sourceMemberAttrs.containsAll(targetMemberAttrs));
+		assertTrue("The source and target member should have the same member-group attributes.",
+			sourceMemberGroupAttrs.containsAll(targetMemberGroupAttrs));
+
+		assertTrue("Ban on VO should be moved from the source member to the target member.",
+			perun.getVosManagerBl().getBanForMember(sess, targetMember.getId()).isPresent());
+
+		assertThatExceptionOfType(MemberNotExistsException.class)
+			.isThrownBy(() -> membersManagerEntry.getMemberById(sess, createdMember.getId()));
+	}
+
+	@Test(expected=AlreadyMemberException.class)
+	public void moveMembershipWhenAlreadyMember() throws Exception {
```

Comment:
Also, these tests don't verify all of the functionality provided by the method. Ideally, you would create a single test to validate a single functionality (e.g. one of coping of attributes, one for group memberships, etc.). However, it would be at least nice to test also the sponsorship copying functionality, even if it should be also part of a single giant test.

To create a single test per single functionality is a good practice because it makes the tests easier to read and also, if some of the tests fail, you know exactly what is wrong. This way, you might need to dig through a lot of code to find out, what caused the problem. However, you don't need to change it here, it is just something to think about in future.

Reply:
I created a single test per single functionality.

Reply:
Well done! :tada: